Add Flux GOVERNANCE doc

scottrigby opened the pull request:
Fixes #1 
Closes #15 

## Intro

This is an initial draft PR for community feedback. @dholbach was kind enough to do an initial quick scan at https://hackmd.io/sBKKaeTgRmO4JYziwq4teQ?both but now it's time to discuss. Please preview at https://github.com/scottrigby/flux-community/blob/governance/GOVERNANCE.md and see what you think 🙂

## Changes for Contributors

### DCO

As soon as this PR is merged, Developer Certificate of Origin (DCO) commit signoff will be required for all new code contributions in the `fluxcd` GitHub org repos.

Note either DCO or CLA is required for CNCF projects. I proposed DCO, as it's easier for end users, and built into git.

For those unfamiliar with DCO, see `git help commit`:

```
-s, --signoff
    Add Signed-off-by line by the committer at the end of the commit log
    message. The meaning of a signoff depends on the project, but it typically
    certifies that committer has the rights to submit this work under the same
    license and agrees to a Developer Certificate of Origin (see
    http://developercertificate.org/ for more information).
```

Your DCO signoff is drawn from your gitconfig's name and email. If you choose to manually add a signoff line, it must be properly formatted and must match your commit information. For example, if you use a GitHub [private email](https://docs.github.com/en/free-pro-team@latest/github/setting-up-and-managing-your-github-user-account/setting-your-commit-email-address) (like `<username>@users.noreply.github.com`) you must make sure to set your gitconfig's email accordingly.

For those who wish to ensure this is always done in your CLI, consider implementing something like [this gist](https://gist.github.com/scottrigby/0c043c0bfbbdb5949e2d824fc3adeaa4). I have successfully used this for several years.

Whenever commit messages are added in the GitHub UI, consider using the [DCO GitHub UI](https://github.com/scottrigby/dco-gh-ui) browser plugin for Chrome or Firefox, which I also maintain, and is in use by a number of contributors to CNCF projects.

## Timeline

- One week for initial comments on this PR (Thurs 15 Oct)
- One week for final thoughts after all initial comments have been addressed (Thurs 22 Oct)
- **Update:** as of Friday 23 Oct there are still details being ironed out. Stay tuned ⏳ 

Readiness to merge the PR will be decided by Lazy consensus. At which point: 
- this GOVERNANCE doc PR will be committed
- we will email the flux-dev list letting contributors know about the DCO change above
- we will enable the DCO bot across all repos in the `fluxcd` GitHub org

We will also notify users and contributors of the above dates, in CNCF Slack #flux channels, and on the flux-dev mailing list.

## Additional Context

**Edit** I removed earlier additional context. I tried to take a less is more approach with the doc itself, since by definition it needs to stand on its own. If interested, see description history for this PR ⏳

hiddeco commented on GOVERNANCE.md:
```diff
@@ -0,0 +1,221 @@
+
+The goal of governing by teams is to allow parts of the project to act independently as appropriate, enabling the project to grow horizontally while remaining well-maintained.
+
+Teams below are also mapped to github teams within the `fluxcd` github org.
+Their members are mapped to github user accounts.
```
```suggestion
Teams below are also mapped to GitHub teams within the `fluxcd` GitHub org.
Their members are mapped to GitHub user accounts.
```

hiddeco commented on GOVERNANCE.md:
```diff
@@ -0,0 +1,221 @@
+
+## Licenses
+
+- Apache 2.0 is required for all source code repositories
+- Developer Certificate of Origin (DCO) commit signoff is required for all new code contributions
```
```suggestion
- Apache 2.0 is required for all source code repositories.
- Developer Certificate of Origin (DCO) commit signoff is required for all new code contributions.
```

hiddeco commented on GOVERNANCE.md:
```diff
@@ -0,0 +1,221 @@
+If that status changes, they will also loose Org team membership.
+- If an Org maintainer volunteraily steps down, in addition to the process above, within 7 calendar days the Flux dev list MUST be notified <https://lists.cncf.io/g/cncf-flux-community>.
+This gives contributors reasonable time to be made aware of the change.
+- When there is an opening for a new Org maintainer, any contributor to a repository in the `fluxcd` github org may nominate a suitable existing member of another team as a replacement.
```
```suggestion
- When there is an opening for a new Org maintainer, any contributor to a repository in the `fluxcd` GitHub org may nominate a suitable existing member of another team as a replacement.
```

hiddeco commented on GOVERNANCE.md:
```diff
@@ -0,0 +1,221 @@
+- Teams oversee activities appropriate to their defined scope and responsibilities.
+- Decisions that affect only one team are made informally by its maintainers.
+- Decisions that affect multiple teams should go through the proposal process.
+- All proposals should be discussed publicly in the appropriate github issue or pull request.
```
```suggestion
- All proposals should be discussed publicly in the appropriate GitHub issue or pull request.
```

hiddeco commented on GOVERNANCE.md:
```diff
@@ -0,0 +1,221 @@
+Pull requests should only be merged after receiving github approval from at least one other member of the appropriate team.
+Voting is not required for most code changes.
+Note that Flux v2 uses github discussions for proposals in the `fluxcd/toolkit` repo <https://github.com/fluxcd/toolkit/discussions?discussions_q=category%3AProposals>.
+- Non-code changes should be proposed as github issues.
```
```suggestion
- Non-code changes should be proposed as GitHub issues.
```

hiddeco commented on GOVERNANCE.md:
```diff
@@ -0,0 +1,221 @@
+## Proposal Process
+
+- Code changes should go through the pull request process, where the idea and implementation details can be publicly discussed with maintainers of the appropriate team, other contributors, and end users.
+Pull requests should only be merged after receiving github approval from at least one other member of the appropriate team.
```
```suggestion
Pull requests should only be merged after receiving GitHub approval from at least one other member of the appropriate team.
```

hiddeco commented on GOVERNANCE.md:
```diff
@@ -0,0 +1,221 @@
+- Code changes should go through the pull request process, where the idea and implementation details can be publicly discussed with maintainers of the appropriate team, other contributors, and end users.
+Pull requests should only be merged after receiving github approval from at least one other member of the appropriate team.
+Voting is not required for most code changes.
+Note that Flux v2 uses github discussions for proposals in the `fluxcd/toolkit` repo <https://github.com/fluxcd/toolkit/discussions?discussions_q=category%3AProposals>.
```
```suggestion
Note that Flux v2 uses GitHub discussions for proposals in the `fluxcd/toolkit` repo <https://github.com/fluxcd/toolkit/discussions?discussions_q=category%3AProposals>.
```

hiddeco reviewed (commented):
Thank you Scott :sun_with_face: 

Did a quick round of review, and on a first read, it all seems reasonable and very much inline with other projects' governance.

hiddeco commented on GOVERNANCE.md:
```diff
@@ -0,0 +1,221 @@
+- release management
+
+In future, the scope and responsibilities of this team may be split into separate dev-related teams.
+But even if so, due to Flux componentized architecture, there should be no assumption of a 1:1 relationship between a source repository and a team.
```
Does this translate to: "one may be a member of the dev team, but only a mainainer of a single Flux project (e.g. `fluxcd/X-controller`, but no others)"?

scottrigby replied:
Ah, no. What does was intended to convey is that even if in future it becomes necessary for the Dev team to split into multiple projects￼ (example: toolkit team, flagger team, helm controller team...), that a git repo does not necessarily equal a "team". For example, a toolkit team would oversee several git repos as part of its scope (fluxcd/toolkit, fluxcd/go-git-providers, etc).

If this feels too clunky we can clarify in some way or just remove it. ￼I added this to preempt confusion because a number of CNCF projects governance assumes a one to one relationship between a git repo and a team or sub-project, but over time or in reality that's not always the case. In the case of Flux, we already know it's not.

hiddeco replied:
Makes sense now that I have read it again, together with your explanation.

What I am missing, given the current state of how Maintainers are already divided across project and how GitHub works with granting permissions to teams, is a section covering "outside maintainers" (as GitHub calls them): people that only maintain a single repository, but do not have rights across the full `fluxcd` GitHub org.

Mainly, because adding them to the Dev team would grant them permissions to:

> - administering, creating, archiving, and deleting source code repositories

which are rights I think we do not want to hand out generously.

Options that come to mind to solve this:
- Already introducing "teams" or sub-projects in the GOVERNANCE doc that have rights scoped to the set of repositories they maintain
- Cover "outside maintainers" in the document, and explain that they are leveled between contributors and the Dev team.
- Make them part of the GitHub org (which gives recognition to people for their efforts, as they can publicly publish it on their profile), but do not assign them to a team and only give them access to the repositories they maintain. This would elevate them from "outside maintainer", while granting them scoped permissions.

squaremo replied:
Yes I think Hidde is right. https://github.com/fluxcd/community/pull/13/files#diff-9e8744ec73212c1966bb19fa22fc3e0aR66 says Maintainers have to serve on one or more teams. I don't think we want to insist on this -- we want some repositories (I am thinking of go-git-providers) to be self-contained and self-sufficient.

scottrigby replied:
OK I have made small changes to address this:
- Moved admin/creating/archiving/deleting git repo responsibility to the Org team
- Clarified a Dev team member may be scoped to one or more git repos, defined by a MAINTAINERS file in each repo (as we do now). Also noted how this is decided (by existing Dev team members listed in a repo's MAINTAINERS file)

WDYT?

scottrigby commented on GOVERNANCE.md:
```diff
@@ -0,0 +1,221 @@
+## Licenses
+
+- Apache 2.0 is required for all source code repositories
+- Developer Certificate of Origin (DCO) commit signoff is required for all new code contributions
```
Just to be clear, this does not refer to PGP signing of commits, but only to DCO git "--signoff￼" on commits.

scottrigby replied:
PS, this browser extension has been useful for helm team members and other people, to add DCO signoff when making commits from the GitHub UI￼ (with her fast PRs that are made UI￼-only, or for committing GitHub PR review "suggestions￼", etc. https://github.com/scottrigby/dco-gh-ui

hiddeco replied:
I am not referring to PGP signatures, but the sign-off using `git commit -s`. See for example the commit history on the `fluxcd/flux` repository as an example that we do not enforce it at present: https://github.com/fluxcd/flux/commits/master

stefanprodan replied:
Please let’s not enforce sign-off on commits, it’s not only completely useless in terms of identity proof  but also a major pain for new contributors when we request them to rebase PRs.

hiddeco replied:
Just to clarify @stefanprodan:

```
       -s, --signoff
           Add Signed-off-by line by the committer at the end of the commit log message. The meaning of a signoff depends on the project, but it typically certifies that committer has the rights to submit this work under
           the same license and agrees to a Developer Certificate of Origin (see http://developercertificate.org/ for more information).
```

I do not see how this would give conflicts while rebasing?

stefanprodan replied:
I’ve seen it many times, when you resolve conflicts, if you keep the last commit message then you’ll be removing the previous signoff. What’s the point of this anyway, without pgp anyone can impersonate another user by just signing off with that user’s email.

hiddeco replied:
> What’s the point of this anyway, without pgp anyone can impersonate another user by just signing off with that user’s email.

It is not about impersonation, but about authorship vs copyright acknowledgement: https://stackoverflow.com/a/1962112

stefanprodan replied:
Ok

caniszczyk commented:
DCO or CLA is also required by CNCF IP policy, you get to choose

On Sun, Oct 4, 2020 at 10:26 AM Hidde Beydals <notifications@github.com>
wrote:

> *@hiddeco* commented on this pull request.
> ------------------------------
>
> In GOVERNANCE.md
> <https://github.com/fluxcd/community/pull/13#discussion_r499258300>:
>
> > +
>
> +Org team membership changes have additional considerations:
>
> +
>
> +- Members of any other team are eligible to be nominiated as an Org team member
>
> +- Org maintainers MUST remain current members of other non-archived teams.
>
> +If that status changes, they will also loose Org team membership.
>
> +- If an Org maintainer volunteraily steps down, in addition to the process above, within 7 calendar days the Flux dev list MUST be notified <https://lists.cncf.io/g/cncf-flux-community>.
>
> +This gives contributors reasonable time to be made aware of the change.
>
> +- When there is an opening for a new Org maintainer, any contributor to a repository in the `fluxcd` github org may nominate a suitable existing member of another team as a replacement.
>
> +  - The nomination period will be three weeks starting the day after an Org team member opening becomes available.
>
> +  - The nomination must be made as a reply to the notification topic in the Flux dev list.
>
> +
>
> +## Licenses
>
> +
>
> +- Apache 2.0 is required for all source code repositories
>
> +- Developer Certificate of Origin (DCO) commit signoff is required for all new code contributions
>
>
> What’s the point of this anyway, without pgp anyone can impersonate
> another user by just signing off with that user’s email.
>
> It is not about impersonation, but about authorship vs copyright
> acknowledgement: https://stackoverflow.com/a/1962112
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/fluxcd/community/pull/13#discussion_r499258300>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAPSIMOXM53S2X646GQ6XLSJCH2NANCNFSM4SCAXW5Q>
> .
>
-- 
Cheers,

Chris Aniszczyk
http://aniszczyk.org
+1 512 961 6719

squaremo commented on GOVERNANCE.md:
```diff
@@ -0,0 +1,221 @@
+
+Org team membership changes have additional considerations:
+
+- Members of any other team are eligible to be nominiated as an Org team member
```
```suggestion
- Members of any other team are eligible to be nominated as an Org team member
```

squaremo replied:
```suggestion
- Members of any other team are eligible to be nominated as an Org team member
```

squaremo commented on GOVERNANCE.md:
```diff
@@ -0,0 +1,221 @@
+
+Org team membership changes have additional considerations:
+
+- Members of any other team are eligible to be nominiated as an Org team member.
```
```suggestion
- Members of any other team are eligible to be nominated as an Org team member.
```

squaremo commented on GOVERNANCE.md:
```diff
@@ -0,0 +1,221 @@
+If that status changes, they will also loose Org team membership.
+- If an Org maintainer volunteraily steps down, in addition to the process above, within 7 calendar days the Flux dev list MUST be notified <https://lists.cncf.io/g/cncf-flux-community>.
+This gives contributors reasonable time to be made aware of the change.
+- When there is an opening for a new Org maintainer, any contributor to a repository in the `fluxcd` GitHub org may nominate a suitable existing member of another team as a replacement.
```
This suggests there's a limited number of Org maintainers, or a process for deciding when there are spots free. That ought to be documented too.

hiddeco replied:
I first thought this too, but later removed my (pending) comment when I came to the conclusion that the interpretation can be very wide. Technically speaking there can be an opening if the current Org maintainers deem it necessary (like a job opening).

However, given we both had this thought, it likely requires some additional documentation :-).

scottrigby replied:
I think you're both right we should clarify this. For example, Helm's Org group states:
> There will be between 3 and 9 people

Do you have any thoughts about what would work best for Flux?

hiddeco replied:
3 as a minimum makes sense to not become authoritarian or end up in a 50/50 conflict. I do not have an exact maximum in mind, except that my experience is that "more isn't always better".

scottrigby replied:
I pushed some changes based on your comments. WDYT?

scottrigby replied:
This is out of date with current simplifications.

squaremo commented on GOVERNANCE.md:
```diff
@@ -0,0 +1,221 @@
+- Members of any other team are eligible to be nominated as an Org team member.
+- Org maintainers MUST remain current members of other non-archived teams.
+If that status changes, they will also loose Org team membership.
+- If an Org maintainer volunteraily steps down, in addition to the process above, within 7 calendar days the Flux dev list MUST be notified <https://lists.cncf.io/g/cncf-flux-community>.
```
```suggestion
- If an Org maintainer voluntarily steps down, in addition to the process above, within 7 calendar days the Flux dev list MUST be notified <https://lists.cncf.io/g/cncf-flux-community>.
```

squaremo commented on GOVERNANCE.md:
```diff
@@ -0,0 +1,221 @@
+Org team membership changes have additional considerations:
+
+- Members of any other team are eligible to be nominated as an Org team member.
+- Org maintainers MUST remain current members of other non-archived teams.
```
How many other?

scottrigby replied:
Good note, we should be more explicit to state 
> Org maintainers MUST remain current members of **at least one** other non-archived team.

squaremo commented on GOVERNANCE.md:
```diff
@@ -0,0 +1,221 @@
+If they are unresponsive for > 3 months they will be automatically removed unless a supermajority of the other members of that team agrees to extend the period.
+- A member may voluntarily step down from a team by following the proposal process in the community repo.
+These proposals are automatically accepted.
+Once offboarding is complete and the issue is closed, they become an emeritus member of that team.
```
I dislike the term "emeritus" in this context -- it's overblown and carries some needless connotations. Do we need to say anything about former team members? I don't think there are any rules relating to that status.

hiddeco replied:
> Do we need to say anything about former team members? I don't think there are any rules relating to that status.

Even if it is not a requirement, I think acknowledging people their past efforts is a nice thing to do, and by making the statement here you guide towards doing this.

What about using a different term, e.g. "retired" (as we [currently do for `fluxcd/flux`](https://github.com/fluxcd/flux/blob/master/MAINTAINERS#L12))?

scottrigby replied:
@squaremo I also adopted this term from Helm, but am not personally set on this academic language. You're right there are no rules around this.

But yes the goal is the same as what @hiddeco said: it's an acknowledgement of past efforts, and encouraging future users to consider helping to do this all too often invisible work.

"retired" seems to fit the same need IMO, and since it's currently used in the project this sounds great! I'll change to use that existing term if there are no objects?

scottrigby replied:
OK changed emeritus -> retired

squaremo commented on GOVERNANCE.md:
```diff
@@ -0,0 +1,221 @@
+
+Scope and responsibilities:
+
+- shared Maintainers responsibilies, as well as
```
```suggestion
- shared Maintainers responsibilities, as well as
```

squaremo reviewed (commented):
Some typo corrections, some comments. I'll batch the typos into a commit, to take them off the radar.

dholbach commented:
CC @michaelbeaumont, @sarataha, @foot, @dinosk, @yiannistri as maintainers of fluxcd/go-git-providers

dholbach commented:
CC @seaneagan as maintainer of fluxcd/helm-controller.

squaremo commented on GOVERNANCE.md:
```diff
@@ -0,0 +1,225 @@
+
+Scope and responsibilities:
+
+- shared Maintainers responsibilies, as well as
```
```suggestion
- shared Maintainers responsibilities, as well as
```

scottrigby replied:
I fixed this with a colon instead. Does this read more clearly now?

scottrigby replied:
haha "responsibilies". This is pretty funny as far as typos go

Thanks, had to fix in separate commit as opposed to committing your suggestion since I had already pushed changes 👍

squaremo commented on GOVERNANCE.md:
```diff
@@ -0,0 +1,225 @@
+Scope and responsibilities:
+
+- shared Maintainers responsibilies, as well as
+- all technical decisions related to project software
```
```suggestion
  all technical decisions related to project software
```

scottrigby replied:
I fixed this with a colon instead. Does this read more clearly now?

squaremo replied:
Yes; I read it as an accidental reformat.

squaremo commented on GOVERNANCE.md:
```diff
@@ -0,0 +1,225 @@
+Once offboarding is complete and the issue is closed, they become a retired member of that team.
+- New members may be added to a team by a supermajority vote by the existing members of that team.
+- When a team has no members the Org team become responsible for it and may either archive the team or find new members.
+- When a team is archived, its members become retired members of that team.
```
When _does_ a team get archived? All of those mentioned seem to be timeless, in the sense of not tied to something with a lifecycle.

scottrigby replied:
There is no defined process as of now. The Org team scope and responsibilities include:
- defining the scope and responsibilities of all other teams

squaremo replied:
Sure, but you could use that answer to justify saying nothing here other than commissioning the Org team, so it's not really an answer. In any case, I think this is referring to _subprojects_ rather than teams, and specifically to what happens to the maintainers when a subproject is archived.

scottrigby replied:
This level of detail has now been removed

seaneagan commented on GOVERNANCE.md:
```diff
@@ -0,0 +1,226 @@
+- Code changes should go through the pull request process, where the idea and implementation details can be publicly discussed with maintainers of the appropriate team, other contributors, and end users.
+Pull requests should only be merged after receiving GitHub approval from at least one other member of the appropriate team.
+Voting is not required for most code changes.
+Note that Flux v2 uses GitHub discussions for proposals in the `fluxcd/toolkit` repo <https://github.com/fluxcd/toolkit/discussions?discussions_q=category%3AProposals>.
```
Should there be any criteria for when a proposal needs a github discussion vs a direct pull request? Perhaps something related to semver e.g. anything which is not a simple PATCH (backward compatible bugfix) change requires a github discussion. Then someone can watch the github discussions to keep up vs needing to watch all repos they care about.

stefanprodan replied:
I would say any change to an existing API that affects end-users should be discussed first, same for new APIs before creating the RFC pull request on a particular repo.

scottrigby replied:
@stefanprodan that sounds right. Do you think we can leave that to be defined by the Architecture team in a separate contribution README, or must we include that detail in this governance document?

stefanprodan replied:
We could have a dedicated doc on this 👍

scottrigby commented on GOVERNANCE.md:
```diff
@@ -0,0 +1,225 @@
+- publishing the vulnerability disclosure once a mitigation strategy is available
+- documenting overall project security best practices
+
+Anyone may email this team privately to safely raise possible security vulnerabilities <https://lists.cncf.io/g/cncf-flux-community>.
```
Yes thanks nice catch. Security team was intended to have its own email. This is critical. It should be set up not as a discussion list, because it must necessarily be a private channel for vulnerability reports.

scottrigby commented on GOVERNANCE.md:
```diff
@@ -0,0 +1,226 @@
+
+- shared Maintainers responsibilities, as well as:
+- all technical decisions related to project software, according to member's scoped git repos
+- release management, according to member's scoped git repos
```
The most recent two commits reflect this 👍 I had been wrestling with whether or not to introduce an "Architecture" team, but Hidde's points solidified the argument for me. Thanks Hidde! The last two co-authored commits reflect the above as well as re-adding Dev team membership considerations, now in a more appropriate location in the doc.

phillebaba commented:
One small thing that I think may be missing is information regarding how a person becomes a maintainer. Is this only possible through invitation or is it something one has to apply to after a certain amount of contributions.

scottrigby commented:
> One small thing that I think may be missing is information regarding how a person becomes a maintainer. Is this only possible through invitation or is it something one has to apply to after a certain amount of contributions.

@phillebaba you're right there's currently only info about how voting takes place, in [Membership Changes](https://github.com/scottrigby/flux-community/blob/governance/GOVERNANCE.md#membership-changes) (for all teams, for an Org team seat, and for maintaining a technical sub-project).

The [Values](https://github.com/scottrigby/flux-community/blob/governance/GOVERNANCE.md#values) section also attempts to expresses that this is an open process, contributions of all kinds from anyone (ideally from a broader spectrum of organizations) may join the community at any level.

I agree there's almost definitely some shared expectations of prior contributions and commitment before nominating/adding a contributor to take on more responsibility within the project. But I didn't want to try adding specific criteria for this to the doc because it seemed better to leave up to the existing members voting process above. I'd like to hear the current maintainers points of view on whether we should detail this out more in the doc or not. I'm trying hard to leverage various project members experiences to keep this both useful and as simple as possible. Please let me know what you think.

squaremo commented on GOVERNANCE.md:
```diff
@@ -0,0 +1,253 @@
+- When a team has no members the Org team become responsible for it and may either archive the team or find new members.
+- When a team is archived, its members become retired members of that team.
+
+Dev team membership changes have additional considerations:
```
Is there anything the Dev team decides that isn't scoped to a project? I reckon either make it explicitly and  _only_ per-project (and exactly the maintainers of each project), or don't treat it as a team alongside the others.

scottrigby replied:
This is out of date with current simplifications.

squaremo reviewed (commented):
Overall comment: this seems like a lot of bureaucracy, while we have not many more maintainers than there are teams proposed. Other comments in line.

squaremo commented on GOVERNANCE.md:
```diff
@@ -0,0 +1,253 @@
+- providing architectural guidance to Dev team members on an individual sub-project
+- representing sub-project interests in Architecture team decisions
+
+Every sub-project should have at least one Architecture team member as a Maintainer.
```
I can see the intent of this, which is a sort of representative democracy; but what I'd expect to happen is a polarisation:

- for some subprojects the maintainers are more or less the architecture team, so a decision about the architecture is really a decision about those subprojects;
- some other subprojects get a passive maintainer, at odds with the the other maintainers, but without any particular stake in the subproject itself.

scottrigby replied:
This has now been removed

squaremo commented:
> this seems like a lot of bureaucracy

More constructively:

 - The Org team has a purpose, as arbiters of last resort (without bothering CNCF, that is)

 - I can see a purpose for the Security team and Community team, distinct from the others.

 - The Dev team doesn't have a remit outside of the individual projects (here I mean the repos under fluxcd/). I think we should just mention the responsibilities of individual repo maintainers and forget about a Dev team.

 - I think the desire for an Architecture team comes from recognising that there's at least one line of development that's spread over several repos (the gitops toolkit bits). But I don't think it would be effective to parachute maintainers into repos so they can be said to be represented, and I don't think it's desirable to have a small group of people deciding how everything works.

It would be better to rely on openness and transparency (i.e., the values) for making sure people get heard -- that should be happening anyway. If there's a bunch of repos that all go together, and a need to co-ordinate, then constitute a team for making decisions over a larger area, but restrict its scope to those repos.

stefanprodan commented:
> The Dev team doesn't have a remit outside of the individual projects (here I mean the repos under fluxcd/). I think we should just mention the responsibilities of individual repo maintainers and forget about a Dev team.

The Dev team would allow repo maintainers to display on their GitHub profile the fluxcd membership. I would keep the Dev team for this purpose only.

> I think the desire for an Architecture team comes from recognising that there's at least one line of development that's spread over several repos (the gitops toolkit bits). But I don't think it would be effective to parachute maintainers into repos so they can be said to be represented, and I don't think it's desirable to have a small group of people deciding how everything works.

I agree, instead of an Architecture team, those that want to get involved into day-to-day decisions they could comment on API proposals and review PRs in each repo.

dholbach commented on GOVERNANCE.md:
```diff
@@ -0,0 +1,253 @@
+
+## Licenses
+
+- Apache 2.0 is required for all source code repositories.
```
I'm wondering if we should explicitly spell out these rules:
- https://github.com/cncf/foundation/blob/master/charter.md#11-ip-policy
- https://github.com/cncf/foundation/blob/master/allowed-third-party-license-policy.md#approved-licenses-for-allowlist
- https://github.com/cncf/foundation/blob/master/copyright-notices.md#copyright-notices

Even if Flux inherits them through being a CNCF project?

scottrigby replied:
I believe this is implied as a CNCF project, and other graduated governance docs don't include it either. However that said, I think this could be noted in a single sentence if we want to include it in this section that we intend to comply with these CNCF policies.

scottrigby replied:
Never mind I like how you added it in your PR

dholbach commented on GOVERNANCE.md:
```diff
@@ -0,0 +1,253 @@
+- engage with end Users through appropriate communication channels appropriate to their teams
+- serve as a point of conflict resolution between Contributors to the Flux area appropriate to their teams
+- maintain open collaboration within and across teams
+- additional responsibilites are defined per team below
```
```suggestion
- additional responsibilities are defined per team below
```

dholbach commented on GOVERNANCE.md:
```diff
@@ -0,0 +1,253 @@
+
+## Teams
+
+The Flux project is governed by Maintainer teams, each with defined scope and resonsibilities.
```
```suggestion
The Flux project is governed by Maintainer teams, each with defined scope and responsibilities.
```

dholbach commented on GOVERNANCE.md:
```diff
@@ -0,0 +1,253 @@
+Architecture team membership changes have additional considerations:
+
+- When an Architecture member leaves the team, their stated maintainership of any technical sub-projects will no longer count toward the required minimum of one Architecture team member.
+  If after leaving, there are no other Architecture team members on those sub-projects, the Architecture team is reponsible for deciding how to fulfill that minimum.
```
```suggestion
  If after leaving, there are no other Architecture team members on those sub-projects, the Architecture team is responsible for deciding how to fulfill that minimum.
```

scottrigby commented:
The most recent commit makes major changes based on the public Flux Dev meeting conversation on this, and follow-up conversation with current Flux maintainers. Thanks @hiddeco for the collaborative editing on this.

The minimal exercise in https://github.com/fluxcd/community/pull/15 was helpful in seeing what we can (and can't) do without 👍  Thanks @dholbach for that.

@stefanprodan your concerns about having a team for maintainers to opt-into displaying membership on their profile is addressed now (we don't speak to this, it's just a feature of the new github team listed). I also hope these changes take into account your points of view from the meeting.

@squaremo I think this addresses your concerns from our feedback chat, from the last Flux meeting, as well as your comments above.

What do we all think about the current state of this PR. Have the controversial parts been addressed? Are we close to being ready to merge this initial document?

squaremo commented on GOVERNANCE.md:
```diff
@@ -0,0 +1,140 @@
+This committee will initially be comprised of Flux Maintainers who have steered the project prior to this initial Governance document.
+The aspiration is no one company or organization should have oversight of the overall project, however that is not yet realistic at this stage. The goal is to broaden maintainership to include a wider range of organizations during CNCF incubation.
+
+Oversight Committee members are publicly listed the `@fluxcd/oversight-committee` <https://github.com/orgs/fluxcd/teams/oversight-committee> GitHub team.
```
```suggestion
Oversight Committee members are publicly listed in the `@fluxcd/oversight-committee` <https://github.com/orgs/fluxcd/teams/oversight-committee> GitHub team.
```

squaremo commented on GOVERNANCE.md:
```diff
@@ -0,0 +1,140 @@
+- If a Maintainer of an affected git repository feels feedback from specific people is warranted they will @mention those users or teams to request feedback.
+- Proposals may also be added to the Flux Dev weekly meetings agenda, as a good avenue for making progress on a decision <https://lists.cncf.io/g/cncf-flux-dev/calendar>.
+- If there are objections and no consensus can be found, a vote may be called by a Maintainer.
+- When a vote is called, Maintainers will cast their yes/no vote on that GitHub issue or pull request, and after a suitable period of time, the outcome will be noted there.
```
Is this any/all maintainers, or maintainers of a specific git repository? If we are going to have rules on voting, we should be clear on who gets to vote. (I would personally rather not have a voting process that includes all maintainers; it's too (easily) weighted towards a single vendor.)

hiddeco replied:
> Decisions that affect only one git repository may be made informally by its Maintainers.

> If there are objections and no consensus can be found, a vote may be called by a Maintainer.

> When a vote is called, Maintainers will cast their yes/no vote on that GitHub issue or pull request, and after a suitable period of time, the outcome will be noted there.

Together imply 

> maintainers of a specific git repository

but I do not think it would hurt to explicitly state this and make it:

> When a vote is called, Maintainers of the git repository will cast their yes/no vote on that GitHub issue or pull request, and after a suitable period of time, the outcome will be noted there.

squaremo replied:
This is better. I think there's still some ambiguity over who gets a vote and who can reasonably say they have a stake. Without assuming malice, I think this could lead to votes being used to bang through decisions, because it's easier and faster than talking it through. So, I would rather have votes limited to the specific situations as set out (i.e., membership and governance). If this turns out to be impractical, we can reconsider.

scottrigby replied:
In the most recent commit, I proposed some very specific clarifications to address this concern.

Slight reorganization for clarity (into "Deciders" and "Kinds of Decisions"). And under Deciders, clarified when the deciders are a specific git repo's Maintainers or the Oversight Committee.

This is distinct from whether a decision can be lazy consensus by the deciders, should go through a proposal process for consensus by the deciders, or put to a vote by the deciders (or if the decision must be escalated).

WDYT?

squaremo replied:
I ought not have understated my objection to voting, in the above.

The strongest kind of agreement is consensus; supermajorities and simple majorities are weaker, even if they feel more official.

From that perspective, only requiring a majority for maintainership, for example, seems rather poor. If you were a maintainer, and you felt strongly enough to go on record voting _against_ someone becoming a maintainer, would you be happy to lose the vote? That does not seem like it would be a good outcome for anyone.

My instinct is to see voting on technical decisions as divisive and ultimately corrosive. But I'd be prepared for individual projects to make their own mind up about that -- maybe the specific people involved are happy voting on low stakes decisions as a way to move fast. For decisions across several projects, I'd argue that voting for the sake of making it go fast is counter-productive.

squaremo replied:
I can see maybe a tie-breaking vote might be required from time to time, but I would hope that it's used as a last resort and that everyone involved agrees a vote is needed (and if they do, they probably already know which way it's going to go -- so it's more a way of signalling that the decision is weaker than a consensus).

To approve this, I would want to see phrasing along the lines of:

 - most day-to-day decisions need only lazy consensus (PRs generally fall into this category, for example)
 - some specific kinds of "high stakes" decision require unanimity

Into the latter category I'd put:

 - election of new maintainers
 - everything currently under "supermajority"

An exception might be things which are high-stakes, but need to be enacted efficiently. Code of conduct violations and removing maintainers might fall here.

In summary: use a _stronger_ standard for governance-level decisions than for day to day decisions; emphasise consensus rather than voting.

scottrigby replied:
@squaremo I've tried to fold these concerns – and the chat with you and @hiddeco – into the most recent commit d52af2c. How is this looking?

squaremo commented on GOVERNANCE.md:
```diff
@@ -0,0 +1,140 @@
+
+## Roles in the Flux Community
+
+The Flux community is comprised of the following roles:
```
```suggestion
The Flux community comprises the following roles:
```

squaremo replied:
Verb rather than adjective is a safer choice for this word

squaremo reviewed (commented):
Just about there. Last area of controversy for me: all-maintainer votes, as noted in a comment.

hiddeco commented on GOVERNANCE.md:
```diff
@@ -0,0 +1,140 @@
+- Enable and promote Flux community values
+- Engage with end Users through appropriate communication channels
+- Serve as a point of conflict resolution between Contributors to their git repository
+- Maintain open collaboration within and across teams
```
```suggestion
- Maintain open collaboration with Maintainers and Contributors
```

(this removes another "team" reference from the document)

hiddeco commented on GOVERNANCE.md:
```diff
@@ -0,0 +1,140 @@
+
+### Simple Majority Changes
+
+- Maintainers: Election of new git project Maintainers by current Maintainers.
```
```suggestion
- Maintainers: Election of new git repository Maintainers by current Maintainers.
```

scottrigby replied:
The 2nd one feels more clear to me 👍

hiddeco commented on GOVERNANCE.md:
```diff
@@ -0,0 +1,140 @@
+- Handling Code of Conduct violations
+- Managing financial decisions
+- Defining the scope of each git repository
+- Resolving escalated decisions when the git repository Maintainers responsible are blocked
```
```suggestion
- Resolving escalated decisions when the Maintainers responsible are blocked
```

I think mentioning "git repository" here is redundant.

scottrigby replied:
Or even just:

```suggestion
- Resolving escalated decisions when Maintainers responsible are blocked
```

stefanprodan reviewed (approved):
LGTM

Thanks @scottrigby and @dholbach for your patience and perseverance 🏅 🍰

hiddeco reviewed (approved):
Thank you for putting this together, your patience, and for the late night discussions @scottrigby :sunflower: :100:

hiddeco commented:
Oh, and @dholbach, thank you for simplifying the draft proposal. Without this it would probably have been a long discussion. :1st_place_medal:

scottrigby commented:
PS there are a number of micro commits on this PR due to me accepting GitHub UI review suggestions. I used this in order to give each person recommending changes their due credit in the commit messages 😄

hiddeco commented:
@scottrigby I would personally smash it  in to a single commit and state all co-authors there.

Including the history in the `main` repository makes little sense in my opinion as what was written there was never enforced or reality.

scottrigby commented:
@hiddeco we're still waiting on @squaremo 's final review of the clarification commit d90ee29. Then we can merge however you prefer.

About merge and squash policies on PRs in general, if it's OK I'd rather not comment too much outside the scope of Governance for this PR. I've made a separate issue for a policy on this https://github.com/fluxcd/community/issues/16 and listed some questions and thoughts around what I've heard so far from the current Flux maintainers.

yiannistri commented:
LGTM thanks to everyone who contributed to this 🥇

foot commented:
Brill indeed, thanks for all your hard work on this doc guys. LGTM.

squaremo commented on GOVERNANCE.md:
```diff
@@ -0,0 +1,159 @@
+- Repository Maintainers: Decisions that affect only one git repository.
+- Oversight Committee: Decisions that are outside the scope of a single git repository.
+
+### Decision Guidelines
```
I <3 the way you've described this Scott -- well articulated!

squaremo commented on GOVERNANCE.md:
```diff
@@ -0,0 +1,159 @@
+## Proposal Process
+
+- Code changes should go through the pull request process, where the idea and implementation details can be publicly discussed with Maintainers, other contributors, and end users.
+  Pull requests should only be merged after receiving GitHub approval from at least one other Maintainer.
```
> after receiving GitHub approval from at least one other Maintainer

If you are not a maintainer, what is the rule?

scottrigby replied:
You're right, the implication is at least one maintainer who is not the PR author. I will make this explicit.

squaremo reviewed (approved):
I am very happy with how this has turned out. Thanks everyone, and especially Scott for carefully navigating the responses.

dholbach reviewed (approved):
Not a maintainer of any fluxcd things, but still very happy with the outcomes of this discussion!

scottrigby commented:
@squaremo My only change since last approvals was 5c23de5.

You can ignore the rest of the force-push. Looks like I had missed a DCO signoff in one of the UI commits from earlier on. Had to fix that now that the DCO bot has been added in #17 (thanks @hiddeco for that!). For future ref, DCO bot gives fix instructions on the checks tab when if fails. In this case it was:
```console
git rebase HEAD~32 --signoff
git push --force-with-lease scottrigby governance
```
